applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py

```diff
@@ -171,35 +190,45 @@ def __deleteAllReportedValues(self):
     # --------------------------------------------------------------------------
     def __isResponseWaitingForInitialValueAsReference(self, response_id):
         response = self.list_of_responses[response_id]
-        if "reference_value" in response and response["reference_value"] == "waiting_for_initial_value":
-            return True
-        else:
-            return False
+        is_reference_defined = "reference" in response
+        if is_reference_defined:
+            is_reference_initial_value = (response["reference"].GetString() == "initial_value")
+            is_reference_value_missing = (response["reference_value"] is None)
+            if is_reference_initial_value and is_reference_value_missing:
+                return True
+        return False
 
     # --------------------------------------------------------------------------
     def __setValueAsReference(self, response_id, value):
         self.list_of_responses[response_id]["reference_value"] = value
 
     # --------------------------------------------------------------------------
     def __translateValueToStandardForm(self, response_id, value):
-        response = self.list_of_responses[response_id]
-        response_type = response["type"]
+        response_type = self.list_of_responses[response_id]["type"]
+        factor = self.list_of_responses[response_id]["scaling_factor"]
+
         if response_type in self.supported_objective_types:
             if response_type == "maximization":
-                return -value
+                return -factor*value
             else:
-                return value
+                return factor*value
         else:
+            reference_value = self.list_of_responses[response_id]["reference_value"]
+
             if response_type == ">" or response_type == ">=":
-                return (response["reference_value"]-value)
+                return factor*(reference_value-value)
             else:
-                return (value-response["reference_value"])
+                return factor*(value-reference_value)
 
     # --------------------------------------------------------------------------
     def __translateGradientToStandardForm(self, response_id, gradient):
         response_type = self.list_of_responses[response_id]["type"]
+        factor = self.list_of_responses[response_id]["scaling_factor"]
+
+        gradient = {key: [factor*value[0],factor*value[1],factor*value[2]] for key, value in gradient.items()}
```

Comment:
dict comprehensions are not the most effective way to simply update all values without using the key information. Here for every key:values pair, a lookup in the dict has to be done (a new dict is created). Instead you can simply iterate over the existing dictionary:

```
for val in gradient.values():
    val[0]*factor
    val[1]*factor
    val[2]*factor
```

I guess the same is valid for many of the dict comprehensions that are used here